Azure/sonic-buildimage
# [Services] Restart BGP service upon unexpected critical process exit

- What I did
Restart BGP service if one of critical processes running in BGP container exited or crashed abnormally.

- How I did it
Generally I follow the framework created by Joe to implement this feature in BGP container.
**First**, add supervisor-proc-exit-listener event listener option in Supervisord configuration file in BGP Docker container.  Supervisord will read a list of critical processes for which to monitor the unexpected crashed and exited.
**Second**, configure bgp.service to always auto-restart the service if it stops, with a delay of 30 seconds. Also set a rate limit of 3 restarts within 20 minutes (1200 seconds).

- How to verify it
On your switch device, please use _docker ps_ command to list all running docker containers.
Then use _docker exec -it container_id /bin/bash_ to login your container. Typing _top_ command
on the shell will display all the processes dynamically and you will spot the process id of one
of the critical processes. Finally type the command _kill -9 process_id_ to kill one process.
After exiting the container, you can use _watch -n 1 docker ps_ to dynamically see the restart
of BGP container.
